### PR TITLE
New Adapter: Rise

### PR DESCRIPTION
**Pull request by misha93m**

### CODE REVIEW

**Comment by gargcreation1992** [comment]
@misha93m PR checks are failing. Requesting you to please check.

**Comment by onkarvhanumante** [comment]
@misha93m  PR is created as Draft. Let us know when PR is ready for review

File: adapters/rise/rise.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+// adapter is a Rise implementation of the adapters.Bidder interface.
+type adapter struct {
+	endpointURL string
+}
+
+func Builder(_ openrtb_ext.BidderName, config config.Adapter, _ config.Server) (adapters.Bidder, error) {
```
Comment (giditietz) [inline adapters/rise/rise.go:22]: why biddername is _?

File: adapters/rise/rise.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	return &adapter{
+		endpointURL: config.Endpoint,
+	}, nil
+}
+
+// MakeRequests prepares the HTTP requests which should be made to fetch bids.
+func (a *adapter) MakeRequests(
+	openRTBRequest *openrtb2.BidRequest,
+	_ *adapters.ExtraRequestInfo) (
```
Comment (giditietz) [inline adapters/rise/rise.go:31]: why ExtraRequestInfo is _?

**Review by giditietz — APPROVED** [review]

**Comment by gargcreation1992** [comment]
@misha93m Please let us know if this PR is ready to be reviewed.

File: adapters/rise/rise.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	requestsToBidder []*adapters.RequestData,
+	errs []error,
+) {
```
Comment (gargcreation1992) [inline adapters/rise/rise.go:34]: Nitpick - for consistency in the codebase please use single line for MakeRequests function declaration. 
```
func (a *adapter) MakeRequests(openRTBRequest *openrtb2.BidRequest, _ *adapters.ExtraRequestInfo) (requestsToBidder []*adapters.RequestData, errs []error) {
//your code
}
```

File: adapters/rise/rise.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+func (a *adapter) extractPublisherID(openRTBRequest *openrtb2.BidRequest) (string, error) {
+	var (
+		bidderExt   adapters.ExtImpBidder
+		impExt      openrtb_ext.ImpExtRise
+		publisherID string
+	)
+
+	for _, imp := range openRTBRequest.Imp {
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err == nil {
```
Comment (gargcreation1992) [inline adapters/rise/rise.go:66]: you should not ignore Unmarshal error here. Please do the error handling properly.
```
if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
     return "", err
}
```

File: adapters/rise/rise.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+			if err = json.Unmarshal(bidderExt.Bidder, &impExt); err == nil && impExt.PublisherID != "" {
```
Comment (gargcreation1992) [inline adapters/rise/rise.go:67]: same comment as above. Please do not ignore Unmarshaling error here.

File: adapters/rise/rise.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	return publisherID, nil
+}
+
+// MakeBids unpacks the server's response into Bids.
+func (a *adapter) MakeBids(
+	request *openrtb2.BidRequest,
+	_ *adapters.RequestData,
+	responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
```
Comment (gargcreation1992) [inline adapters/rise/rise.go:84]: Requesting you to please move this to single line for consistency across the codebase.

File: adapters/rise/rise.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
```
Comment (gargcreation1992) [inline adapters/rise/rise.go:101]: Checking response status code this way is correct but you could also make use of the following code - 
```
if httputil.IsResponseStatusCodeNoContent(response) {
		return nil, nil
	}

	if err := httputil.CheckResponseStatusCodeForErrors(response); err != nil {
		return nil, []error{err}
	}
```

File: adapters/rise/rise.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+func getMediaTypeForBid(impID string, imps []openrtb2.Imp) openrtb_ext.BidType {
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Banner != nil {
+				return openrtb_ext.BidTypeBanner
+			} else if imp.Video != nil {
+				return openrtb_ext.BidTypeVideo
+			}
```
Comment (gargcreation1992) [inline adapters/rise/rise.go:130]: Warning - This is an anti-pattern. This code assumes that for multi-format bid request, mediatype will be set to type banner by default. I would recommend to set the mediaType based on bidder response. Your bidder should be able to set openrtb2.Bid.Mtype field and use that field here to get mediaTypeForImp. Even if your bidder does not support multi-format request, I would highly encourage you to make this change. [Here is an example.](https://github.com/prebid/prebid-server/blob/6b616196605030f568dae4a727d1a658de929417/adapters/ix/ix.go#L246-L247)

File: static/bidder-info/rise.yaml [hunk 9]
```diff
@@ -0,0 +1,18 @@
+endpoint: "https://pbs.yellowblue.io/pbs"
+maintainer:
+  email: rise-prog-dev@risecodes.com
```
Comment (gargcreation1992) [inline static/bidder-info/rise.yaml:3]: Sent an email confirmation to verify email addresss. Please reply received.

Comment (gargcreation1992) [reply static/bidder-info/rise.yaml:3]: email verified. thanks!

File: static/bidder-info/rise.yaml [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    mediaTypes:
+      - banner
+      - video
+  site:
+    mediaTypes:
+      - banner
+      - video
+userSync:
+  iframe:
+    url: https://pbs-cs.yellowblue.io/pbs-iframe?gdpr={{.GDPR}}&gdpr_consent={{.GDPRConsent}}&us_privacy={{.USPrivacy}}&redirect={{.RedirectURL}}
```
Comment (gargcreation1992) [inline static/bidder-info/rise.yaml:17]: User sync URL works. 👍

**Review by gargcreation1992 — CHANGES_REQUESTED** [review]
Requesting few changes.

File: adapters/rise/rise.go [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+	return append(requestsToBidder, &adapters.RequestData{
+		Method:  http.MethodPost,
+		Uri:     a.endpointURL + "?publisher_id=" + publisherID,
+		Body:    openRTBRequestJSON,
+		Headers: headers,
+	}), nil
+}
+
+func (a *adapter) extractPublisherID(openRTBRequest *openrtb2.BidRequest) (string, error) {
```
Comment (onkarvhanumante) [inline adapters/rise/rise.go:58]: `a *adapter` is not being referred inside function body.  Therefore no need to define `extractPublisherID` on `(a *adapter)`

**Comment by onkarvhanumante** [comment]
@misha93m  tests are failing with following error:
```
Error: adapters/rise/rise.go:55:14: undefined: httputil.IsResponseStatusCodeNoContent
Error: adapters/rise/rise.go:59:21: undefined: httputil.CheckResponseStatusCodeForErrors
```
`IsResponseStatusCodeNoContent` and `CheckResponseStatusCodeForErrors` were moved to `adapters` package. Could you sync PR branch with master branch updates

File: adapters/rise/rise.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+	switch bid.MType {
+	case openrtb2.MarkupBanner:
+		return openrtb_ext.BidTypeBanner, nil
+	case openrtb2.MarkupVideo:
+		return openrtb_ext.BidTypeVideo, nil
+	case openrtb2.MarkupAudio:
+		return openrtb_ext.BidTypeAudio, nil
+	case openrtb2.MarkupNative:
+		return openrtb_ext.BidTypeNative, nil
+	default:
```
Comment (gargcreation1992) [inline adapters/rise/rise.go:122]: In rise.yaml, you have defined that your bidder will be supporting just Banner and Video. I believe you might wanna error out if bidder response returns Audio or Native as bid type. So here is the code - 
```
switch bid.MType {
	case openrtb2.MarkupBanner:
		return openrtb_ext.BidTypeBanner, nil
	case openrtb2.MarkupVideo:
		return openrtb_ext.BidTypeVideo, nil
	default:
                 return "", fmt.Errorf("unsupported MType %d", bid.MType)
```

**Review by onkarvhanumante — APPROVED** [review]

**Review by gargcreation1992 — APPROVED** [review]

**Review by gargcreation1992 — APPROVED** [review]

**Pull request merged**